shared/src/balancer/graph_api.rs

```diff
@@ -37,41 +39,47 @@ impl BalancerSubgraphClient {
     }
 
     /// Retrieves the list of registered pools from the subgraph.
-    pub async fn get_weighted_pools(&self) -> Result<(u64, Vec<RegisteredWeightedPool>)> {
+    pub async fn get_weighted_pools(&self) -> Result<RegisteredWeightedPools> {
         let block_number = self.get_safe_block().await?;
-        let mut results = Vec::<RegisteredWeightedPool>::new();
-        while {
-            // We do paging by last ID instead of using `skip`. This is the
-            // suggested approach to paging best performance:
-            // <https://thegraph.com/docs/graphql-api#pagination>
-            let last_id = results
-                .last()
-                .map(|pool| json!(pool.pool_id))
-                .unwrap_or_else(|| json!(""));
+        let mut pools_by_factory = HashMap::<H160, Vec<RegisteredWeightedPool>>::new();
 
+        // We do paging by last ID instead of using `skip`. This is the
+        // suggested approach to paging best performance:
+        // <https://thegraph.com/docs/graphql-api#pagination>
+        let mut last_id = H256::default();
+        while {
             let page = self
                 .0
                 .query::<pools_query::Data>(
                     pools_query::QUERY,
                     Some(json_map! {
                         "block" => block_number,
                         "pageSize" => QUERY_PAGE_SIZE,
-                        "lastId" => last_id,
+                        "lastId" => json!(last_id),
                     }),
                 )
                 .await?
                 .pools;
+
             let has_next_page = page.len() == QUERY_PAGE_SIZE;
+            if let Some(last_pool) = page.last() {
+                last_id = last_pool.id;
+            }
 
-            results.reserve(page.len());
             for pool in page {
-                results.push(pool.into_registered(block_number)?);
+                pools_by_factory
+                    .entry(pool.factory.unwrap_or_default())
```

Comment:
but default here will fallback on the zero address... How will we know what to do with this if we get it?

Reply:
My reasoning here is that when we get all the pools, any "factory buckets" that aren't one of the factories we are expecing (WeightedPoolFactory or Weighted2TokenPoolFactory) will trigger an allert - including the 0x00...00 address.